src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexOpcode.cs

```diff
@@ -0,0 +1,162 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.Text.RegularExpressions
+{
+    /// <summary>Opcodes written by <see cref="RegexWriter"/> and used by <see cref="RegexInterpreter"/> to process a regex.</summary>
+    /// <remarks>
+    /// <see cref="RegexCode"/> stores an int[] containing all of the codes that make up the instructions for
+    /// the interpreter to process the regular expression.  The array contains a packed sequence of operations,
+    /// each of which is an <see cref="RegexOpcode"/> stored as an int, followed immediately by all of the operands
+    /// required for that operation.  For example, the subexpression `a{2,7}[^b]` would be represented as the sequence
+    ///     0 97 2 3 97 5 10 98
+    /// which is interpreted as:
+    ///     0  = opcode for Onerep (a{2, 7} is written out as a repeater for the minimum followed by a loop for the maximum minus the minimum)
+    ///     97 = 'a'
+    ///     2  = repeat count
+    ///     3  = opcode for Oneloop
+    ///     97 = 'a'
+    ///     5  = max iteration count
+    ///     10 = opcode for Notone
+    ///     98 = 'b'
+    /// </remarks>
+    internal enum RegexOpcode
+    {
+        // Primitive operations
+
+        /// <summary>Repeater of the specified character.</summary>
+        /// <remarks>Operand 0 is the character. Operand 1 is the repetition count.</remarks>
+        Onerep = 0,
+        /// <summary>Repeater of a single character other than the one specified.</summary>
+        /// <remarks>Operand 0 is the character. Operand 1 is the repetition count.</remarks>
+        Notonerep = 1,
+        /// <summary>Repeater of a single character matching the specified set</summary>
+        /// <remarks>Operand 0 is index into the strings table of the character class description. Operand 1 is the repetition count.</remarks>
+        Setrep = 2,
+
+        /// <summary>Greedy loop of the specified character.</summary>
+        /// <remarks>Operand 0 is the character. Operand 1 is the max iteration count.</remarks>
+        Oneloop = 3,
+        /// <summary>Greedy loop of a single character other than the one specified.</summary>
+        /// <remarks>Operand 0 is the character. Operand 1 is the max iteration count.</remarks>
+        Notoneloop = 4,
+        /// <summary>Greedy loop of a single character matching the specified set</summary>
+        /// <remarks>Operand 0 is index into the strings table of the character class description. Operand 1 is the repetition count.</remarks>
+        Setloop = 5,
+
+        /// <summary>Lazy loop of the specified character.</summary>
```

Comment:
is there value in keeping the little samples? eg this one was `.{,n}?`

Reply:
I don't recall, how are non zero based loops represented? Is `a{1,3}` represented as OneRep followed by OneLoop of 2?

Reply:
> is there value in keeping the little samples? eg this one was .{,n}?

We're pretty far removed here.  These are the result of nodes being transformed by the writer into opcodes, so I'm not sure that replication is helpful, and it's also a bit misleading in some cases because of transformations like...

> how are non zero based loops represented? Is a{1,3} represented as OneRep followed by OneLoop of 2?

Exactly.
https://github.com/dotnet/runtime/pull/64346/files/04a301237597ff45d1e721894c48f462f45d3de1#diff-ff01da6b61635cc68262dc66ccf774c5dc742cbbbdb9349abf27203bbefc8448R14